Logstash/logstash-input-azurewadtable/lib/logstash/inputs/azurewadtable.rb

```diff
@@ -167,8 +183,154 @@ def partitionkey_from_datetime(time_string)
 
   # Convert time to ticks
   def to_ticks(time_to_convert)
-    @logger.debug("Converting time to ticks")
+    #@logger.debug("Converting time to ticks")
     time_to_convert.to_i * 10000000 - TICKS_SINCE_EPOCH
   end # to_ticks
 
 end # LogStash::Inputs::AzureWADTable
+
+class AzureQuery
+  def initialize(logger, azure_table_service, table_name, query_str, query_id, entity_count_to_process)
+    @logger = logger
+    @query_str = query_str
+    @query_id = query_id
+    @entity_count_to_process = entity_count_to_process
+    @azure_table_service = azure_table_service
+    @table_name = table_name
+    @continuation_token = nil
+  end
+
+  def reset
+    @continuation_token = nil
+  end
+
+  def id
+    return @query_id
+  end
+
+  def run(on_result_cbk)
+    results_found = false
+    @logger.debug("[#{@query_id}]Query filter: " + @query_str)
+    begin
+      @logger.debug("[#{@query_id}]Running query. continuation_token: #{@continuation_token}")
+      query = { :top => @entity_count_to_process, :filter => @query_str, :continuation_token => @continuation_token }
+      result = @azure_table_service.query_entities(@table_name, query)
+
+      if result and result.length > 0
+        results_found = true
+        @logger.debug("[#{@query_id}] #{result.length} results found.")
+        result.each do |entity|
+          on_result_cbk.call(entity)
+        end
+      end
+
+      @continuation_token = result.continuation_token
+    end until !@continuation_token
+
+    return results_found
+  end
+end
+
+class QueryData
+  def initialize(logger, query)
+    @logger = logger
+    @query = query
+    @results_cache = Set.new
+  end
+
+  def id
+    return @query.id
+  end
+
+  def get_unique_id(entity)
+    uniqueId = ""
+    partitionKey = entity.properties["PartitionKey"]
+    rowKey = entity.properties["RowKey"]
+    uniqueId << partitionKey << rowKey
+    return uniqueId
+  end
+
+  def run_query
+    results = []
+    @query.reset
+    @query.run( ->(entity) {
+      uniqueId = get_unique_id(entity)
+
+      if @results_cache.add?(uniqueId).nil?
+        @logger.debug("[#{@query.id}][QueryData] #{uniqueId} already processed")
+      else
+        @logger.debug("[#{@query.id}][QueryData] #{uniqueId} new item")
+        results.push(entity)
+      end
+    })
+    return results
+  end
+
+  def has_entity(entity)
+    return @results_cache.include?(get_unique_id(entity))
+  end
+
+end
+
+class DuplicateDetector
+  def initialize(logger, past_queries_count)
+    @logger = logger
+    @past_queries_count = past_queries_count
+    @query_cache = []
```

Comment:
Have you considered use hashtable to hold the cache? It will reduce complicity from O(n) to O(1) if implement correctly and you won't need the loop to fetch do the result comparison on your own: https://ruby-doc.org/core-2.2.0/Hash.html.

Reply:
I chose an array of sets because I can easily invalidate the oldest query. For a hash implementation I would have to iterate through the entire structure each time and the complexity would be O(m) where m is the number of results (far greater than the number of queries)